### ISSUE TITLE
Suggestion: better error message with JavetProxyConverter

### ISSUE THREAD

**Issue by ceclin** [caoccao/Javet#131]
See [this](https://github.com/caoccao/Javet/blob/4d57f9063a3d0f52ad96a0aaa7050d073cf1e82c/src/main/java/com/caoccao/javet/interop/proxy/JavetUniversalProxyHandler.java#L1002). Since you use reflection, Throwable `t` is an InvocationTargetException in most cases, whose `getMessage()` is always null. 

I think it is better to use `t.getCause().getMessage()` or add a specific catch for InvocationTargetException.

In some cases, user cannot get the root cause Throwable, such as IJavetPromiseRejectCallback. I think it is better to replace `getMessage()` with `toString()`. The latter includes class name of Throwable, which provides more useful message.

**Comment by caoccao** [comment]
Thank you for reporting this issue. Please let me know if the fix works or not.

**Comment by caoccao** [comment]
With the new [commit](https://github.com/caoccao/Javet/commit/106a9c9059a7f677af0162a4165b30869b1a1a6c), the inner exception is more accurate. Please take a try and let me know if it works.

**Comment by ceclin** [comment]
Sorry. I am too busy to test it in last two weeks.

It works well with my code. Thank you!

**Comment by ceclin** [comment]
But I still have some questions. For example, with JavetProxyConverter:
```java
runtime.getGlobalObject().set("Java", new Object() {
    public Class<?> type(String className) throws ClassNotFoundException {
        return Class.forName(className);
    }
});
```
```java
try {
    runtime.getExecutor(script).setResourceName(name).executeVoid();
} catch (BaseJavetScriptingException e) {
    JavetScriptingError scriptingError = e.getScriptingError();
    runtime.getLogger().error("failed to execute: " + scriptingError.getStack());
}
```
```js
Java.type('class that does not exists');
```
Is there a way to get the root cause (ClassNotFoundException) in the catch block?

**Comment by caoccao** [comment]
Yes, the inner exception is the one (ClassNotFoundException). It was recently fixed. Please make sure you are using the latest snapshot [build](https://github.com/caoccao/Javet/actions).

**Comment by ceclin** [comment]
I just downloaded [linux build for f17de21](https://github.com/caoccao/Javet/actions/runs/1636332182). I am sure that `e.getCause()` returns null.

**Comment by caoccao** [comment]
Thank you for the info. It was a bug and just got fixed. Please try the latest build and let me know if you have any questions.

**Comment by ceclin** [comment]
It works. Thank you!

**Comment by ceclin** [comment]
POC:
```java
try (V8Runtime runtime = V8Host.getV8Instance().createV8Runtime()) {
    runtime.setConverter(new JavetProxyConverter());
    runtime.setPromiseRejectCallback((event, promise, value) -> {
        System.err.println(value);
    });
    JavetStandardConsoleInterceptor console = new JavetStandardConsoleInterceptor(runtime);
    V8ValueGlobalObject global = runtime.getGlobalObject();
    console.register(global);
    runtime.getGlobalObject().set("Class", Class.class);
    try (V8Module module = runtime.getExecutor("Class.forName('abc');")
            .setResourceName("module.js")
            .setModule(true)
            .compileV8Module()) {
        module.executeVoid();
    }
    runtime.getGlobalObject().delete("Class");
    console.unregister(global);
    runtime.lowMemoryNotification();
} catch (Exception e) {
    System.err.println(e);
}
```
If it uses `compileV8Module().executeVoid()` like above, both PromiseRejectCallback and catch block handle the exception.
If it simply uses `executeVoid()`, only PromiseRejectCallback handles the exception, which I think is expected.

**Comment by caoccao** [comment]
Nice catch. A routine was missed. This [commit](https://github.com/caoccao/Javet/commit/235f790e791ba39b696b705793702ac1976c5e5f) should work as you expected. Please take a try.

**Comment by ceclin** [comment]
Yes, it works.

**Comment by ceclin** [comment]
If a java method throws an exception in js async function, the result Promise is rejected with V8ValueError, whose error message is `callback {name} failed with error message {e.getMessage()}`. In this case, V8ValueError doesn't provide any way to get the cause Exception. Is it easy to implement? If no, would you consider replacing `e.getMessage()` with `e.toString()` or something like `e.getClass().getSimpleName() + e.getMessage()`. Class name of Exception is often useful or even essential information, such as NullPointerException (`Objects.requireNonNull`), FileNotFoundException (HttpURLConnection 404) and so on.

**Comment by caoccao** [comment]
Could you provide the reproduceable code and expected output?

`toString()` sometimes is too much for JS to understand. The coding experience in JS needs to be considered as well. Exposing Java class info in V8 is information leakage, I think.

**Comment by ceclin** [comment]
```java
public class POC {
    public static void main(String[] args) throws Exception {
        try (V8Runtime runtime = V8Host.getV8Instance().createV8Runtime()) {
            runtime.setConverter(new JavetProxyConverter());
            runtime.setPromiseRejectCallback((event, promise, value) -> {
                System.err.println(value);
            });
            JavetStandardConsoleInterceptor console = new JavetStandardConsoleInterceptor(runtime);
            V8ValueGlobalObject global = runtime.getGlobalObject();
            console.register(global);
            POC poc = new POC();
            runtime.getGlobalObject().bind(poc);
            runtime.getExecutor("request('https://cdn.jsdelivr.net/npm/jqery/dist/jquery.min.js');")
                    .setResourceName("module.js")
                    .setModule(true)
                    .executeVoid();
            runtime.getGlobalObject().unbind(poc);
            console.unregister(global);
            runtime.lowMemoryNotification();
        }
    }

    @V8Function
    public String request(String url) throws IOException {
        InputStream inputStream = new URL(url).openStream();
        ByteArrayOutputStream result = new ByteArrayOutputStream();
        byte[] buffer = new byte[4096];
        for (int length; (length = inputStream.read(buffer)) != -1; )
            result.write(buffer, 0, length);
        return result.toString("UTF-8");
    }
}
```
Output:
```
Error: https://cdn.jsdelivr.net/npm/jqery/dist/jquery.min.js
```
Maybe better:
```
Error: FileNotFoundException: https://cdn.jsdelivr.net/npm/jqery/dist/jquery.min.js
```

**Comment by ceclin** [comment]
Another example:
```java
public class POC {
    public static void main(String[] args) throws Exception {
        try (V8Runtime runtime = V8Host.getV8Instance().createV8Runtime()) {
            runtime.setConverter(new JavetProxyConverter());
            runtime.setPromiseRejectCallback((event, promise, value) -> {
                System.err.println(value);
            });
            JavetStandardConsoleInterceptor console = new JavetStandardConsoleInterceptor(runtime);
            V8ValueGlobalObject global = runtime.getGlobalObject();
            console.register(global);
            runtime.getGlobalObject().set("poc", POC.class);
            runtime.getExecutor("poc.foo(null);")
                    .setResourceName("module.js")
                    .setModule(true)
                    .executeVoid();
            runtime.getGlobalObject().delete("poc");
            console.unregister(global);
            runtime.lowMemoryNotification();
        }
    }

    public static void foo(Object obj) {
        Objects.requireNonNull(obj);
    }
}
```
Output:
```
Error: Callback foo failed with error message <null>
```
Maybe better:
```
Error: Callback foo failed with error message NullPointerException
```

**Comment by caoccao** [comment]
## Example 1

I think there are 2 options.

1. Catch the `IOException` and throw whatever you want to throw with customized message.
2. Javet may enhance the `Error` object to carry the inner exception in JS. That could be a nice feature in the future.

## Example 2

If you try `Objects.requireNonNull(o, "The argument is null.");` in Java, you will get `Error: The argument is null.` in JS.

## Summary

Javet aims at being as less intrusive as possible. The current behavior allows applications to control everything to be exposed in JS to achieve maximal information security. Please let me know if you have any questions.

**Comment by ceclin** [comment]
> Catch the `IOException` and throw whatever you want to throw with customized message.

> If you try `Objects.requireNonNull(o, "the argument is null.");` in Java, you will get `Error: Callback foo failed with error message the argument is null.` in JS.

In fact, I am developing a library based on Javet. It is a pity that I can't control everything. I will tell my library users that they can better error message in this way.

> Javet may enhance the `Error` object to carry the inner exception in JS. That could be a nice feature in the future.

I would appreciate it if this feature is implemented.

------

Thank you for solving these issues!

By the way, when will 1.0.7 be released?

**Comment by caoccao** [comment]
Is your library open-sourced? The enhancement for `Error` will be after v1.0.7 is released. When? Chrome 97 will be released soon (Jan 4), so will v1.0.7 be.

**Comment by ceclin** [comment]
It is not open-sourced for now, but it is on the plan.
